pallets/anchors-v2/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet_prelude::*;
+use frame_support::traits::ReservableCurrency;
+use frame_system::pallet_prelude::*;
+use scale_info::TypeInfo;
+use sp_runtime::{traits::AtLeast32BitUnsigned, FixedPointOperand};
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::*;
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct Anchor<AccountId, DocumentId, DocumentVersion, Hash, Balance> {
+	account_id: AccountId,
+	document_id: DocumentId,
+	document_version: DocumentVersion,
+	hash: Hash,
+	deposit: Balance,
+}
+
+pub type AnchorOf<T> = Anchor<
+	<T as frame_system::Config>::AccountId,
+	<T as Config>::DocumentId,
+	<T as Config>::DocumentVersion,
+	<T as frame_system::Config>::Hash,
+	<T as Config>::Balance,
+>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type Balance: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ From<u128>
+			+ TypeInfo
+			+ TryInto<u64>;
+
+		type Currency: ReservableCurrency<Self::AccountId, Balance = Self::Balance>;
+
+		/// Default deposit that will be taken when adding an anchor.
+		type DefaultAnchorDeposit: Get<Self::Balance>;
+
+		/// The type used to identify a document.
+		type DocumentId: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ From<u128>
+			+ TypeInfo
+			+ TryInto<u64>;
+
+		/// The type used to version a document.
+		type DocumentVersion: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ TypeInfo
+			+ TryInto<u64>;
+
+		/// Origin used when setting a deposit.
+		type AdminOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Weight information.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Storage for document anchors.
+	#[pallet::storage]
+	#[pallet::getter(fn get_anchor)]
+	pub type Anchors<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_256,
+		(T::DocumentId, T::DocumentVersion),
+		Blake2_256,
+		T::AccountId,
+		AnchorOf<T>,
+	>;
+
+	/// Storage for document anchors specific to an account.
+	#[pallet::storage]
+	#[pallet::getter(fn get_personal_anchor)]
+	pub type PersonalAnchors<T: Config> = StorageNMap<
+		_,
+		(
+			NMapKey<Blake2_256, T::AccountId>,
+			NMapKey<Blake2_256, T::DocumentId>,
+			NMapKey<Blake2_256, T::DocumentVersion>,
+		),
+		AnchorOf<T>,
+		OptionQuery,
+	>;
```

Comment:
To avoid storing the anchor twice (which leads to maintaining the same data in two storages), I would create a simple `StorageMap<AnchorId, Anchor>` and would store in these two maps the `AnchorId` instead of the `Anchor`.

You would have another Read operation but the main data is stored just once.

Reply:
Another option, because the keys of both maps are the same, then if the anchor is found in `PersonalAnchors,` you know that it can also be found in `Anchors` So `PersonalAnchors` could have `()` as a value. Then if `Some(())`, then you can retrieve the anchor data in `Anchors`.

Reply:
Thank you for the suggestions, I will definitely explore these and get back to you!

Reply:
We had a similar approach initially in the `pallet_order_book`, and later we changed it because it did not scale well. You can check `Orders` and `UserOrders` storages for reference